python/taichi/types/ndarray_type.py

```diff
@@ -9,78 +12,107 @@ def __init__(self, element_type, shape=None):
-        # TODO Replace CompoundType with MatrixType and VectorType
-
-        if isinstance(dtype, CompoundType):
-            if dtype == TensorType:
-                raise TypeError(
-                    "TensorType is not supported for ndarray dtype annotation."
-                )
-            self.element_dim = dtype.ndim
-            self.element_shape = (dtype.n, ) if dtype.ndim == 1 else (dtype.n,
-                                                                      dtype.m)
-            if not (self.element_dim > 0 and self.element_dim <= 2):
-                raise TypeError(
-                    f"Unexpected matrix data type {dtype} has dimension {dtype.ndim}, only vectors and matrices (ndim = 1,2) are accepted."
-                )
+        # TODO(Haidong) Deprecate element shape in 1.4.0. Use dtype to manage element-level arguments.
+        if element_dim is not None or element_shape is not None:
+            warnings.warn(
+                "The element_dim and element_shape arguments for ndarray will be deprecated in v1.4.0, use matrix dtype instead.",
+                DeprecationWarning)
+            self.dtype = make_matrix_dtype_from_element_shape(
+                element_dim, element_shape, dtype)
         else:
-            self.element_shape = element_shape
-            self.element_dim = element_dim
+            self.dtype = dtype
 
         self.field_dim = field_dim
         self.layout = Layout.AOS
 
     def check_matched(self, ndarray_type: NdarrayTypeMetadata):
-        if self.element_dim is not None and self.element_dim > 0:
-            if not isinstance(ndarray_type.element_type, TensorType):
-                raise TypeError(
-                    f"Expect TensorType element for Ndarray with element_dim: {self.element_dim} > 0"
-                )
-            if self.element_dim != len(ndarray_type.element_type.shape()):
-                raise ValueError(
-                    f"Invalid argument into ti.types.ndarray() - required element_dim={self.element_dim}, but {len(ndarray_type.element_type.shape())} is provided"
-                )
-
-        if self.element_shape is not None and len(self.element_shape) > 0:
-            if not isinstance(ndarray_type.element_type, TensorType):
-                raise TypeError(
-                    f"Expect TensorType element for Ndarray with element_shape: {self.element_shape}"
-                )
-
-            if self.element_shape != ndarray_type.element_type.shape():
-                raise ValueError(
-                    f"Invalid argument into ti.types.ndarray() - required element_shape={self.element_shape}, but {ndarray_type.element_type.shape()} is provided"
-                )
+        # FIXME(Haidong) Cannot use Vector/MatrixType due to circular import
+        # Use the CompuoundType instead to determine the specific typs.
+        # TODO Replace CompoundType with MatrixType and VectorType
+        if isinstance(self.dtype, CompoundType):
```

Comment:
nit: can we just check if `self.dtype` and `ndarray_type.dtype` match instead of translating them back to element_dim and element_shape? 
This may look like:
```
if isinstance(self.dtype, CompoundType):
   assert isinstance(ndarray_type.dtype, CompoundType)
   # check self.dtype.element_shape, shape and layout matches
else:
  assert isinstance(ndarray_type.dtype, primitive_dtypes) # pseudo code
  # check self.dtype matches ndarray_type.dtype
```

Reply:
Exact match of `self.dtype` and `ndarray_type` is too strict for our use cases at the moment. There are `dtype` of MatrixType with `None` shape and `None` dim and `None` type. So the refactored match code is no cleaner than this bunch of if-else. I did the translation just to keep it backward compatible, and yes, we should clean up the logics for type matching for ndarrays. I've get rid of the `element_shape` and `element_dim` vars, but I think it's better to do a full sweep in a follow-up PR, WDYT?